--- trafficserver-8.0.5.orig/proxy/http/HttpConfig.h	2019-08-20 16:58:34.000000000 +0000
+++ trafficserver-8.0.5/proxy/http/HttpConfig.h	2019-09-13 00:13:28.596134237 +0000
@@ -782,7 +782,10 @@
   enum {
     CACHE_REQUIRED_HEADERS_NONE                   = 0,
     CACHE_REQUIRED_HEADERS_AT_LEAST_LAST_MODIFIED = 1,
-    CACHE_REQUIRED_HEADERS_CACHE_CONTROL          = 2
+    CACHE_REQUIRED_HEADERS_CACHE_CONTROL          = 2,
+    // NOTE: This is a value added by my private patch, so choose a value which is unlikely
+    // to added in the traffciserver upstream.
+    CACHE_REQUIRED_HEADERS_CACHE_CONTROL_S_MAXAGE = 99
   };
 
   enum {
--- trafficserver-8.0.5.orig/proxy/http/HttpTransact.cc	2019-08-20 16:58:34.000000000 +0000
+++ trafficserver-8.0.5/proxy/http/HttpTransact.cc	2019-09-13 00:13:28.600134152 +0000
@@ -6060,6 +6060,14 @@
         }
         break;
 
+      case HttpConfigParams::CACHE_REQUIRED_HEADERS_CACHE_CONTROL_S_MAXAGE:
+        if (!(response->get_cooked_cc_mask() & MIME_COOKED_MASK_CC_S_MAXAGE)) {
+          TxnDebug("http_trans", "[is_response_cacheable] "
+                                 "s-maxage is required");
+          return false;
+        }
+        break;
+
       default:
         break;
       }
